Support for the program data address space option of LLVM's Target Datalayout

TimNN opened the pull request:
This was introduced recently (specifically, for AVR, cc @dylanmckay).

(I came up with this when attempting to run [avr-rust](https://github.com/avr-rust/rust) rebased on the latest [rust-lang](https://github.com/rust-lang/rust) commits. If this requires a different design, some additional discussions, or is not something to pursue right now, I'd be happy to close this PR).

Note that this somewhat overlaps with @DiamondLovesYou's #51576, I think, although the implementation here is significantly simpler: Since the address space applies to _all_ program data, we can just check the pointee's type whenever we create an LLVM pointer type. If it is a function we use the program data address space; if not we use the default address space.

cc @eddyb, who has been reviewing #51576

Ref: https://llvm.org/docs/LangRef.html#data-layout

rust-highfive commented:
r? @nikomatsakis

(rust_highfive has picked a reviewer for you, use r? to override)

rust-highfive commented:
The job `x86_64-gnu-llvm-5.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/440215272) ([raw log](https://api.travis-ci.org/v3/job/440215272/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain

[00:04:44] travis_fold:start:tidy
travis_time:start:tidy
tidy check
[00:04:44] tidy error: /checkout/src/librustc_target/abi/mod.rs:109: line longer than 100 chars
[00:04:45] some tidy checks failed
[00:04:45] 
[00:04:45] 
[00:04:45] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/tidy" "/checkout/src" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "--no-vendor" "--quiet"
[00:04:45] 
[00:04:45] 
[00:04:45] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test src/tools/tidy
[00:04:45] Build completed unsuccessfully in 0:00:47
[00:04:45] Build completed unsuccessfully in 0:00:47
[00:04:45] make: *** [tidy] Error 1
[00:04:45] Makefile:79: recipe for target 'tidy' failed

The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:0cdda034
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
---
travis_time:end:02bd36e4:start=1539273876699334755,finish=1539273876704878228,duration=5543473
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:01dbbf09
$ ln -s . checkout && for CORE in obj/cores/core.*; do EXE=$(echo $CORE | sed 's|obj/cores/core\.[0-9]*\.!checkout!\(.*\)|\1|;y|!|/|'); if [ -f "$EXE" ]; then printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" "$CORE"; gdb --batch -q -c "$CORE" "$EXE" -iex 'set auto-load off' -iex 'dir src/' -iex 'set sysroot .' -ex bt -ex q; echo travis_fold":"end:crashlog; fi; done || true
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:0531b499
travis_time:start:0531b499
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No such file or directory
travis_fold:end:after_failure.5
travis_fold:start:after_failure.6
travis_time:start:19dc590e
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

nikomatsakis commented on src/librustc_codegen_llvm/type_.rs:
```diff
@@ -233,7 +234,26 @@ impl Type {
+        if self.kind() == TypeKind::Function {
+            unsafe {
+                llvm::LLVMPointerType(self,
+                                      cx.data_layout().program_memory_address_space as c_uint)
```
It is possible to "mis-use" this, I assume, for example via transmutes or something? Can we wind up with "valid Rust code" giving weird LLVM assertion failures or other errors?

TimNN replied:
That's a good point. I'll see what happens if I transmute a few things to a function pointer.

nikomatsakis commented:
cc @rust-lang/compiler — seems like an ad-hoc extension, nominating for brief T-compiler discussion to decide best path

TimNN commented:
> seems like an ad-hoc extension

Yeah, this is definitely a bit ad-hoc, however overall the approach doesn't seem too bad to me. In any case, I came up with this while trying to get a target to work that isn't even officially supported yet. I'd be totally happy if you said this requires some more design work in the future.

nikomatsakis commented:
@TimNN to be clear, I didn't mean "ad hoc" in the pejorative sense. Just that it's one of those "small features" that we often decide without an RFC -- it might however warrant a feature gate or a tracking issue or some such thing.

DiamondLovesYou commented:
My patch can be easily extended to support this. But I think the space should be called the instruction address space, since "program memory" is kinda ambiguous (even though this is the term LLVM uses).

@eddyb wanted any value to be cast into the flat address space before that value is used (ie uses of a `global` would actually use an `addrspacecast`-ed const expr), in order to reduce the patch's complexity footprint. @TimNN would this be possible for the AVR target machine? The initial version I've whipped up doesn't force this backflip (it uses the correct space for functions when creating LLVM types from our types), but @eddyb might want them to go through a flat space cast round trip.

TimNN commented:
>  think the space should be called the instruction address space

I'd be fine with that, as you noticed I was just copying LLVM terminology.

> @eddyb wanted any value to be cast into the flat address space before that value is used (ie uses of a `global` would actually use an `addrspacecast`-ed const expr), in order to reduce the patch's complexity footprint. @TimNN would this be possible for the AVR target machine?

I don't know. I understand how my patch works but overall I'm mostly unfamiliar with the involved concepts. IIRC, however, LLVM was unhappy casting a function to a function pointer in an incorrect addrspace (In case that information helps you).

---

In any case, if you think that the functionality added here could be better supported with your patch, I'd be happy to wait for that to be merged. (I'd probably open this PR against the `avr-rust` repo for others to use in the meantime, but for me it is in no way urgent to get this submitted).

bors commented:
:umbrella: The latest upstream changes (presumably #55171) made this pull request unmergeable. Please resolve the merge conflicts.

eddyb commented on src/librustc_codegen_llvm/type_.rs:
```diff
@@ -233,7 +234,26 @@ impl Type {
 
-    pub fn ptr_to(&self) -> &Type {
+    pub fn ptr_to(&self, cx: &CodegenCx<'ll, '_>) -> &Type {
+        if self.kind() == TypeKind::Function {
```
I really want to avoid us introspecting LLVM types. I think we should either not use non-flat address space pointers *or* we should special-case `fn` pointers - from a quick search, I found only 2 places that create them:
https://github.com/rust-lang/rust/blob/40123a1257f7645bee1c6d3b6eb8c2f6eebd206c/src/librustc_codegen_llvm/type_of.rs#L268
https://github.com/rust-lang/rust/blob/40123a1257f7645bee1c6d3b6eb8c2f6eebd206c/src/librustc_codegen_llvm/meth.rs#L42

Since these are both calling `.ptr_to()` on the result of `FnType::llvm_type`, a helper method on `FnType` can probably be added.

eddyb commented on src/librustc_target/abi/mod.rs:
```diff
@@ -96,6 +106,8 @@ impl TargetDataLayout {
             match spec.split(':').collect::<Vec<_>>()[..] {
                 ["e"] => dl.endian = Endian::Little,
                 ["E"] => dl.endian = Endian::Big,
+                [ref p] if p.starts_with("P") =>
```
```suggestion
                [p] if p.starts_with("P") =>
```

DiamondLovesYou commented:
@TimNN Well, I've push a change to support the feature needed in this PR, so feel free to give it a try.

> I don't know. I understand how my patch works but overall I'm mostly unfamiliar with the involved concepts. IIRC, however, LLVM was unhappy casting a function to a function pointer in an incorrect addrspace (In case that information helps you).

It does help. As I said, my patch put the vtable entries and all declared functions in the instruction address space, so that should work. I just realized it would not, however, do the same for function ptrs cast from non fn ptr values, like integers (perhaps a bit of an edge case). Currently, it tries to cast the value to the flat address space just like any other ptr.

But, baring that, it should work for AVR with a suitable target spec. If you'd like to give my patch a shot, let me know how it goes!

I would definitely prefer to have my patch merged instead; it (now) supports (that int -> fn ptr issue not withstanding) this address space, and can compile Rust which then runs on an AMDGPU.

TimNN commented:
Thanks for your reviews, @eddyb & @nikomatsakis! 

I'm not sure what you think, but I feel like @DiamondLovesYou's PR #51576 is the way to go in the long term. 

I however don't know what the timeline would be to get that merged (or if you maybe have concerns that would prevent merging it). 

I see two ways forward:

1) close this for now until a decision on #51576 is reached.
2) merge this for now (after addressing your comments) and refactor / upgrade things in #51576.

What do you all think?

shepmaster commented:
A drive-by comment from AVR-land...

We think that users would like / need the ability to annotate certain pieces of memory with which address space they should go in. We opened [an issue](https://github.com/avr-rust/rust/issues/84) about this, but in short, something like this strawman syntax could address the issue:

```rust
#[space(progmem)]
pub static FOOBAR: &'static str = "hello world";
```

I don't know that this PR would handle something like that, but it may be worth keeping in mind for the underlying design.

TimNN commented:
I've updated this PR to address the review comments. However I'll close this for now until there's a decision on ##51576.

rust-highfive commented:
Your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/448459575) ([raw log](https://api.travis-ci.org/v3/job/448459575/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain

```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

eddyb commented:
> I'm not sure what you think, but I feel like @DiamondLovesYou's PR #51576 is the way to go in the long term.

This is a much smaller change, and I'd rather include this now, before we can figure out #51576.

eddyb commented:
r=me with rebase

pnkfelix commented:
Visited for T-compiler triage. This no longer appears to be blocked on waiting for our team.

TimNN commented:
@bors r=eddyb

bors commented:
:pushpin: Commit 52b7a714bbe0cc2385b7df609cd1b7569f98f2f0 has been approved by `eddyb`

<!-- @bors r=eddyb 52b7a714bbe0cc2385b7df609cd1b7569f98f2f0 -->

rust-highfive commented:
The job `x86_64-gnu-llvm-5.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/452466971) ([raw log](https://api.travis-ci.org/v3/job/452466971/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_time:end:187b845a:start=1541694066672287627,finish=1541694067889269200,duration=1216981573
$ git checkout -qf FETCH_HEAD
travis_fold:end:git.checkout

Encrypted environment variables have been removed for security reasons.
See https://docs.travis-ci.com/user/pull-requests/#Pull-Requests-and-Security-Restrictions
$ export SCCACHE_BUCKET=rust-lang-ci-sccache2
$ export SCCACHE_REGION=us-west-1
Setting environment variables from .travis.yml
$ export IMAGE=x86_64-gnu-llvm-5.0
---

[00:04:10] travis_fold:start:tidy
travis_time:start:tidy
tidy check
[00:04:10] tidy error: /checkout/src/librustc_target/abi/mod.rs:109: line longer than 100 chars
[00:04:11] some tidy checks failed
[00:04:11] 
[00:04:11] 
[00:04:11] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/tidy" "/checkout/src" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "--no-vendor" "--quiet"
[00:04:11] 
[00:04:11] 
[00:04:11] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test src/tools/tidy
[00:04:11] Build completed unsuccessfully in 0:00:46
[00:04:11] Build completed unsuccessfully in 0:00:46
[00:04:11] Makefile:79: recipe for target 'tidy' failed
[00:04:11] make: *** [tidy] Error 1

The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:01a234be
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
---
travis_time:end:00201a56:start=1541694330129382526,finish=1541694330134786665,duration=5404139
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:13b29031
$ ln -s . checkout && for CORE in obj/cores/core.*; do EXE=$(echo $CORE | sed 's|obj/cores/core\.[0-9]*\.!checkout!\(.*\)|\1|;y|!|/|'); if [ -f "$EXE" ]; then printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" "$CORE"; gdb --batch -q -c "$CORE" "$EXE" -iex 'set auto-load off' -iex 'dir src/' -iex 'set sysroot .' -ex bt -ex q; echo travis_fold":"end:crashlog; fi; done || true
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:3352d47f
travis_time:start:3352d47f
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No such file or directory
travis_fold:end:after_failure.5
travis_fold:start:after_failure.6
travis_time:start:140d18f0
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

TimNN commented:
@bors r- (forgot to run tidy)

TimNN commented:
@bors r=eddyb

bors commented:
:pushpin: Commit 50a2d47b4f57f8f9433b18692fe2d1975f9e84e1 has been approved by `eddyb`

<!-- @bors r=eddyb 50a2d47b4f57f8f9433b18692fe2d1975f9e84e1 -->

bors commented:
:hourglass: Testing commit 50a2d47b4f57f8f9433b18692fe2d1975f9e84e1 with merge f705d13c0de0f6721468e867ae993bbf819ddbbd...

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/builds/20162621)

kennytm commented:
@bors retry 

3 hour timeout in `x86_64-msvc`, likely due to LLVM cache purge.

TimNN commented:
@bors retry

* Another timeout on AppVeyor due to LLVM rebuilding?

bors commented:
:hourglass: Testing commit 50a2d47b4f57f8f9433b18692fe2d1975f9e84e1 with merge 9b8f9029762da90a88c8ca6f8ff7690177ec696a...
